sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -200,14 +200,20 @@ object ConstantPropagation extends Rule[LogicalPlan] {
 
   private def replaceConstants(condition: Expression, equalityPredicates: EqualityPredicates)
     : Expression = {
-    val constantsMap = AttributeMap(equalityPredicates.map(_._1))
-    val predicates = equalityPredicates.map(_._2).toSet
-    def replaceConstants0(expression: Expression) = expression transform {
+    val allConstantsMap = AttributeMap(equalityPredicates.map(_._1))
+    val allPredicates = equalityPredicates.map(_._2).toSet
+    def replaceConstants0(
+        expression: Expression, constantsMap: AttributeMap[Literal]) = expression transform {
       case a: AttributeReference => constantsMap.getOrElse(a, a)
     }
     condition transform {
-      case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
-      case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
+      case b: BinaryComparison =>
```

Comment:
To support other binary comparisons. For example: `>`, `<`.

Reply:
Can we look at the commit history and understand why constant propagation only supports equality predicates before?

Reply:
The initial commit did not support it. It seems other binary comparisons were not considered at the time:
https://github.com/apache/spark/pull/17993

Later we tried to enhance it. It seems it is not easy because there are too many changes:
https://github.com/apache/spark/pull/24553

Reply:
cc @peter-toth

Reply:
I think we can replace constants in other expressions too, https://github.com/apache/spark/pull/24553 does the same (and more).

Reply:
btw, @wangyum, @cloud-fan, I can revive https://github.com/apache/spark/pull/24553 if we want to enhance constant propagation a bit more...

Reply:
It seems https://github.com/apache/spark/pull/24553 is a more comprehensive optimization, but we should keep the algorithm simple:
1. Collecting constants for non-attribute expressions seems very complicated to me. What if the non-attribute expressions have dependencies by themselves?
2. We should avoid replacing constants in expressions recursively. To support `a = 1 AND a = 2` ===> `false`, can we detect it earlier when building the attribute -> constant map and return false?

Reply:
1. I'm not sure I get the issue. If a complex expression is equal to a constant why can't we use that constant where the complex expression appears? I don't think it matter if some parts of the complex expression can also be replaced to another constat. The replace happens using `transformDown` so larger complex expressions are replaced to constants first. E.g. `a = 1 AND a + b = 1 AND a + b > 2` -> `1 + b = 1 AND 1 > 2` -> `false` makes sense to me.

3. Yes we can. The reason why I didn't do those kind of shortcuts in `ConstantPropagation` in my PR is because there are other, similar cases that are not that easy to shortcut (or have been handled in other rules so why do it here as well) . E.g. `a = 1 AND a > 2` -> `false` where `1 > 2` could be also handled in `ConstantPropagation` but those foldable expressions are already handled in `ConstantFolding` and then in `BooleanSimplification` when contant propagation has been done.

Reply:
Let me open a PR tomorrow.

Reply:
I've opened https://github.com/apache/spark/pull/40268.
I didn't want to overcomplicate the change so took only a few parts from https://github.com/apache/spark/pull/24553. But if you think 1. or 2. or any other feature from the original PR is also needed I can add it to the PR.

Reply:
@cloud-fan This is why conflicting equality predicates are not supported: https://github.com/apache/spark/pull/17993#discussion_r117619380
It is correct if it only optimize on the filter condition.

Reply:
Sidenote: In my old PR (https://github.com/apache/spark/pull/24553/files#diff-d43484d56a4d9991066b5c00d12ec2465c75131e055fc02ee7fb6dfd45b5006fR76-R79) I proposed  https://github.com/apache/spark/pull/27309 to detect equijoins better and to allow constant propagation in `Join`s too.